8330814: Cleanups for KeepAliveCache tests

RealCLanger opened the pull request:
While working in that area I found some potential for cleanup of a few tests.

Most notably:

B5045306.java:
- does not need to run in othervm
- the executor service that it uses should be shut down eventually to free resources

B8291637.java:
- use just one instead of two test VM invocations
KeepAliveTimerThread.java:
call to grp.destroy() at the end is pointless (API is void & deprecated for removal)

Generally:
The deprecated URL constructor is used. It can be switched to the handy URIBuilder
Some more try with resources here and there

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330814](https://bugs.openjdk.org/browse/JDK-8330814): Cleanups for KeepAliveCache tests (**Bug** - P4)


### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [3296795b](https://git.openjdk.org/jdk/pull/18884/files/3296795b61c955bb8eb8899455c7a109e55ddf0c)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18884/head:pull/18884` \
`$ git checkout pull/18884`

Update a local copy of the PR: \
`$ git checkout pull/18884` \
`$ git pull https://git.openjdk.org/jdk.git pull/18884/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18884`

View PR using the GUI difftool: \
`$ git pr show -t 18884`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18884.diff">https://git.openjdk.org/jdk/pull/18884.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18884#issuecomment-2069127751)

turbanoff commented on test/jdk/sun/net/www/http/KeepAliveCache/B5045306.java:
```diff
@@ -63,150 +66,166 @@
+            long[] threads = threadMXBean.getAllThreadIds();
+            ThreadInfo[] threadInfo = threadMXBean.getThreadInfo(threads);
+            for (int i=0; i<threadInfo.length; i++) {
+                if (threadInfo[i].getThreadName().equals("Keep-Alive-SocketCleaner"))  {
```
```suggestion
                if (threadInfo[i].getThreadName().equals("Keep-Alive-SocketCleaner")) {
```

jaikiran commented on test/jdk/sun/net/www/http/KeepAliveCache/B8291637.java:
```diff
@@ -125,20 +130,26 @@ public void run() {
+                .loopback()
+                .port(port)
+                .path("/firstCall")
+                .toURL();
```
Hello Christoph, what's the significance of changing the path from `/` to `/firstCall` in this test?

RealCLanger replied:
Good catch, copy&paste error.

jaikiran commented on test/jdk/sun/net/www/http/KeepAliveCache/B8291637.java:
```diff
@@ -125,20 +130,26 @@ public void run() {
+        try (InputStream i = urlc.getInputStream()) {
+            int c;
+            byte[] buf = new byte[256];
+            while ((c=i.read(buf)) != -1) {
```
Nit: needs a space before and after the `=`

jaikiran commented on test/jdk/sun/net/www/http/KeepAliveCache/B8291637.java:
```diff
@@ -125,20 +130,26 @@ public void run() {
+            int c;
+            byte[] buf = new byte[256];
+            while ((c=i.read(buf)) != -1) {
+                count+=c;
```
Same here, needs a space before/after `+=`

RealCLanger replied:
Fixed

jaikiran commented on test/jdk/sun/net/www/http/KeepAliveCache/B5045306.java:
```diff
@@ -42,14 +42,17 @@
+
 /* Part 1:
- * The http client makes a connection to a URL whos content contains a lot of
+ * The http client makes a connection to a URL who's content contains a lot of
```
I think it should be "whose"

jaikiran commented on test/jdk/sun/net/www/http/KeepAliveCache/B5045306.java:
```diff
@@ -63,150 +66,166 @@
-
-            //Only read 1 byte of response data and close the stream
-            InputStream is = uc.getInputStream();
+        System.out.println("http server listens on: " + server.getAddress().getPort());
```
Maybe move this log message to the line after where we call server.start() and perhaps print the `server.getAddress()` instead of just the port.

RealCLanger replied:
Done

jaikiran commented on test/jdk/sun/net/www/http/KeepAliveCache/B5045306.java:
```diff
@@ -63,150 +66,166 @@
 
     public static void startHttpServer() {
         try {
             server = HttpServer.create(new InetSocketAddress(InetAddress.getLocalHost(), 0), 10, "/", new SimpleHttpTransactionHandler());
```
While we are changing this test, can you change this to `InetAddress.getLoopbackAddress()` and then when constructing the requet URI, use `URIBuilder.loopback()`?

RealCLanger replied:
Done

jaikiran commented on test/jdk/sun/net/www/http/KeepAliveCache/B5045306.java:
```diff
@@ -63,150 +66,166 @@
+                        break;  // thread has died, OK
+
+                    // if Keep-Alive-SocketCleaner consumes more than 50% of cpu then we
+                    // can assume a recursive loop.
```
Interesting test case. I'm a bit surprised we haven't seen this intermittently fail. Speaking of which, I see that the change to this test proposes to remove it from `/othervm` and potentially run it in agent vm mode. Given the kind of checks this test is doing, I think it's better to let it run as a `othervm` test to keep the test as isolated as possible.

RealCLanger replied:
I've thought about this, too. However, I see the only critical point why it could merit a `/othervm` test is this thing with querying the thread CPU time of the Keep-Alive-SocketCleaner thread. But I think the likelihood of this failing is the same within standard test and `/othervm`. So I'd prefer to change to standard test, since this is one of the improvements of this change.

dfuch replied:
The cache is global. Who knows what it might already contain when the test starts and what may be left behind when the test ends. So I agree that it may be prudent to keep `/othervm`

RealCLanger replied:
OK, then reverting back to `/othervm`.

dfuch commented on test/jdk/sun/net/www/http/KeepAliveCache/B8291637.java:
```diff
@@ -125,20 +130,26 @@ public void run() {
-        i.close();
-        System.out.println("Read " + count );
+        System.out.println("Read " + count);
         try {
```
Maybe the `try {` line should be moved just after `server.start()`

RealCLanger replied:
good point. I enhanced this a little bit more. 😄

RealCLanger commented:
ping (@dfuch) - any updates/approvals 😄

jaikiran commented on test/jdk/sun/net/www/http/KeepAliveCache/B8291637.java:
```diff
@@ -125,28 +127,35 @@ public void run() {
+            try (InputStream i = urlc.getInputStream()) {
+                int c;
+                byte[] buf = new byte[256];
+                while ((c = i.read(buf)) != -1) {
```
Given that the server response is just a hello world message, I think you can just replace this entire code within the try block with something like:

```java
int count;
try (InputStream i = urlc.getInputStream()) {
    count = i.readAllBytes().length;
}
```

If you prefer to leave it in the current form, that's fine too.

RealCLanger replied:
Done.

jaikiran reviewed (approved):
Hello Christoph, the latest changes look fine to me. I've added a trivial comment about reading the inputstream in one of the test, but you can leave it in its current form if you prefer to.
Please run `tier2` tests once before integrating.

RealCLanger commented:
> Hello Christoph, the latest changes look fine to me. I've added a trivial comment about reading the inputstream in one of the test, but you can leave it in its current form if you prefer to. Please run `tier2` tests once before integrating.

Thanks for the review. I added your suggestion. Will run it once more through our test suite tonight and then integrate tomorrow if all is good.

dfuch reviewed (commented):
Looks reasonable. Give us some time to test that in our CI.

RealCLanger commented:
> Looks reasonable. Give us some time to test that in our CI.

Our CI testing looks good. Please let me know when you are done.

RealCLanger commented:
> Looks reasonable. Give us some time to test that in our CI.

@dfuch May I merge now?

RealCLanger commented:
If I don't hear back, I'll integrate tomorrow. Should we see problems later on, I'm available for following up. 😄

dfuch commented:
@RealCLanger sorry about that: I ran the tests then got distracted and forgot to come back. You're good to go - no issue found!

RealCLanger commented:
> @RealCLanger sorry about that: I ran the tests then got distracted and forgot to come back. You're good to go - no issue found!

Cool. Thanks for the confirmation.

/integrate
